### PR TITLE
vl53l0x: Add support for interrupt and configurable sampling frequency

### PR DESCRIPTION
**Pull request by titouanc**
- drivers: sensor: vl53l0x: run clang-format
- drivers: sensor: vl53l0x: add attr_get/set for the sampling frequency
- drivers: sensor: vl53l0x: add support for interrupt-gpios
- boards: shields: x_nucleo_53l0a1: add interrupt-gpios
- samples: sensors: vl53l0x: update to use trigger mode if available

### CODE REVIEW

File: drivers/sensor/vl53l0x/vl53l0x.c [hunk 1]
```diff
@@ -263,9 +263,79 @@ static int vl53l0x_channel_get(const struct device *dev, enum sensor_channel cha
 	return 0;
 }
 
+static int vl53l0x_get_sampling_freq(const struct device *dev, struct sensor_value *val)
+{
+	uint32_t timing;
+	struct vl53l0x_data *drv_data = dev->data;
+	int ret = VL53L0X_GetMeasurementTimingBudgetMicroSeconds(&drv_data->vl53l0x, &timing);
+	if (ret) {
+		LOG_ERR("[%s] VL53L0X_GetMeasurementTimingBudgetMicroSeconds failed", dev->name);
```
Comment (MaureenHelm) [inline drivers/sensor/vl53l0x/vl53l0x.c:377]: Please use more meaningful error messages instead of just printing painfully long function names.

File: drivers/sensor/vl53l0x/vl53l0x.c [hunk 2]
```diff
@@ -263,9 +263,79 @@ static int vl53l0x_channel_get(const struct device *dev, enum sensor_channel cha
...
+		return ret;
+	}
+	val->val1 = 1000000 / timing;
+	val->val2 = 1000000 / (1000000 % timing) / timing;
```
Comment (MaureenHelm) [inline drivers/sensor/vl53l0x/vl53l0x.c:381]: Can timing ever be zero?

File: drivers/sensor/vl53l0x/vl53l0x.c [hunk 3]
```diff
@@ -263,9 +263,79 @@ static int vl53l0x_channel_get(const struct device *dev, enum sensor_channel cha
...
+	return ret;
+}
+
+static int vl53l0x_set_sampling_freq(const struct device *dev, const struct sensor_value *val)
+{
+	uint64_t timing = 1000000000000 / (1000000 * val->val1 + val->val2);
```
Comment (MaureenHelm) [inline drivers/sensor/vl53l0x/vl53l0x.c:387]: Use `USEC_PER_SEC` and friends

File: drivers/sensor/vl53l0x/vl53l0x.c [hunk 4]
```diff
@@ -42,16 +43,77 @@ LOG_MODULE_REGISTER(VL53L0X, CONFIG_SENSOR_LOG_LEVEL);
...
+	struct gpio_dt_spec interrupt;
 	struct gpio_dt_spec xshut;
 };
 
 struct vl53l0x_data {
+	const struct device *dev;
 	bool started;
 	VL53L0X_Dev_t vl53l0x;
-	VL53L0X_RangingMeasurementData_t RangingMeasurementData;
+	VL53L0X_RangingMeasurementData_t measurement;
```
Comment (MaureenHelm) [inline drivers/sensor/vl53l0x/vl53l0x.c:54]: This change doesn't look related to interrupt-gpios. Please extract to a separate commit.

File: drivers/sensor/vl53l0x/vl53l0x.c [hunk 5]
```diff
@@ -42,16 +43,77 @@ LOG_MODULE_REGISTER(VL53L0X, CONFIG_SENSOR_LOG_LEVEL);
...
+{
+	switch (chan) {
+	case SENSOR_CHAN_DISTANCE:
+	case SENSOR_CHAN_PROX:
+	case SENSOR_CHAN_ALL:
+		return true;
+	default:
+		return false;
+	}
+}
```
Comment (MaureenHelm) [inline drivers/sensor/vl53l0x/vl53l0x.c:72]: This change doesn't look related to interrupt-gpios. Please extract to a separate commit.

File: drivers/sensor/vl53l0x/vl53l0x.c [hunk 6]
```diff
@@ -140,6 +196,18 @@ static int vl53l0x_setup_single_shot(const struct device *dev)
...
+{
+	struct vl53l0x_data *drv_data = dev->data;
+	int ret = VL53L0X_SetDeviceMode(&drv_data->vl53l0x, mode);
+	if (ret) {
+		LOG_ERR("[%s] VL53L0X_SetDeviceMode failed", dev->name);
+		return ret;
+	}
+	drv_data->current_mode = mode;
+	return 0;
+}
```
Comment (MaureenHelm) [inline drivers/sensor/vl53l0x/vl53l0x.c:209]: This change doesn't look related to interrupt-gpios. Please extract to a separate commit.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/vl53l0x/vl53l0x.c (hunk 1 above)
Comment (titouanc) [reply drivers/sensor/vl53l0x/vl53l0x.c:377]: I agree with that. However all the existing log messages in this driver (since its birth in Zephyr https://github.com/zephyrproject-rtos/zephyr/commit/347560459b0c55efe95e2764197a96ce22247b02) use this same form, therefore we would have to change them all for consistency ?

Comment (MaureenHelm) [reply drivers/sensor/vl53l0x/vl53l0x.c:377]: If you're up for that yes, but I'd also be fine with just making your new log messages meaningful even if they're inconsistent with the old ones.

File: drivers/sensor/vl53l0x/vl53l0x.c (hunk 6 above)
Comment (titouanc) [reply drivers/sensor/vl53l0x/vl53l0x.c:209]: In my opinion, this change is related to interrupt gpios. Before this commit, the sensor was always in single-shot mode (ie the i2c master tells the sensor to start a measurement, waits, then fetch the result from the sensor).

In this commit, we add support for `sensor_trigger_set` for `SENSOR_TRIG_DATA_READY` in the vl53l0x driver. For this trigger to make sense, we need to put the sensor in continuous mode (ie the sensor continuously makes new measurements, and notifies the i2c master via an interrupt that a new measurement is ready).

In continuous mode, the caller does not need to start a new measurement in `sensor_sample_fetch`, so we need to track the current mode of the device to know if a new measurement should be started. This function simply wraps the call to the ST HAL, so that the current device mode is properly tracked. 

Finally, the device mode is not exposed to the user, so this function is only an accompanying helper for the interrupt gpio functionality

**Comment by titouanc** [comment]
Thank you for reviewing this PR @MaureenHelm 

I applied your suggestions, fixed some issues uncovered by the compliance checks, and rebased on top of the latest `main`

**Comment by titouanc** [comment]
I clicked on "Re-request review" for @MaureenHelm and this removed all the other reviewers, not sure why :thinking:

File: drivers/sensor/vl53l0x/vl53l0x.c [hunk 7]
```diff
@@ -62,52 +62,40 @@ static int vl53l0x_setup_single_shot(const struct device *dev)
 
 	ret = VL53L0X_StaticInit(&drv_data->vl53l0x);
 	if (ret) {
-		LOG_ERR("[%s] VL53L0X_StaticInit failed",
```
Comment (jfischer-no) [inline drivers/sensor/vl53l0x/vl53l0x.c:65]: Don't, unnecessary changes, https://github.com/zephyrproject-rtos/zephyr/issues/48785

**Review by jfischer-no — CHANGES_REQUESTED** [review]
